comp/dogstatsd/server/parse.go

```diff
@@ -206,9 +201,17 @@ func (p *parser) parseMetricSample(message []byte) (dogstatsdMetricSample, error
 				return dogstatsdMetricSample{}, fmt.Errorf("dogstatsd timestamp should be > 0")
 			}
 			timestamp = time.Unix(ts, 0)
-		// container ID
+		// local data
 		case p.dsdOriginEnabled && bytes.HasPrefix(optionalField, localDataPrefix):
-			containerID = p.resolveContainerIDFromLocalData(optionalField)
+			rawLocalData := string(optionalField[len(localDataPrefix):])
+			localData, err = origindetection.ParseLocalData(rawLocalData)
+			if err != nil {
+				return dogstatsdMetricSample{}, fmt.Errorf("failed to parse OriginInfo.LocalData %s: %v", rawLocalData, err)
+			}
+			// If the container ID is not set in the Local Data, we try to resolve it from the cgroupv2 inode.
+			if localData.ContainerID == "" {
+				localData.ContainerID = p.resolveContainerIDFromInode(localData.Inode)
+			}
```

Comment:
This snippet is repeated for every data type in dogstatsd. Would it be possible to move it into a helper?

Also, this introduces a new error path where there was none before - invalid local data didn't prevent the metrics and events from progressing down the pipeline, but does now. Is it intended? If yes, it deserves a mention in the changelog. Either way, we probably want to test that this works one way or the other, could you please add a test?